pennylane_lightning/src/tests/Test_GateImplementations_CompareKernels.cpp

```diff
@@ -0,0 +1,186 @@
+ * @param num_qubits Number of qubits
+ * @param wires Wires the gate applies to
+ * @param inverse Whether to use inverse of gate
+ * @param params Paramters for gate
+ */
+template <Gates::GateOperation gate_op, typename PrecisionT, typename ParamT,
+          typename GateImplementation, class Alloc>
+auto applyGate(std::vector<std::complex<PrecisionT>, Alloc> ini,
+               size_t num_qubits, const std::vector<size_t> &wires,
+               bool inverse, const std::vector<ParamT> &params)
+    -> std::vector<std::complex<PrecisionT>, Alloc> {
+    callGateOps(GateOpToMemberFuncPtr<PrecisionT, ParamT, GateImplementation,
+                                      gate_op>::value,
+                ini.data(), num_qubits, wires, inverse, params);
+    return ini;
+}
+
+/**
+ * @brief Apply the given gate using all implementing kernels and return
+ * results in tuple.
+ */
+template <Gates::GateOperation gate_op, typename PrecisionT, typename ParamT,
+          typename Kernels, class Alloc, size_t... I>
+auto applyGateForImplemetingKernels(
+    const std::vector<std::complex<PrecisionT>, Alloc> &ini, size_t num_qubits,
+    const std::vector<size_t> &wires, bool inverse,
+    const std::vector<ParamT> &params,
+    [[maybe_unused]] std::index_sequence<I...> dummy) {
+    return std::make_tuple(
+        applyGate<gate_op, PrecisionT, ParamT, getNthType<Kernels, I>>(
+            ini, num_qubits, wires, inverse, params)...);
+}
+
+template <Gates::GateOperation gate_op, typename PrecisionT, typename ParamT,
+          class RandomEngine>
+void testApplyGate(RandomEngine &re, size_t num_qubits) {
+    const auto ini = createRandomState<PrecisionT>(re, num_qubits);
+
+    using Kernels = typename KernelsImplementingGate<gate_op>::Type;
+
+    INFO("Kernels implementing " << lookup(gate_names, gate_op) << " are "
+                                 << kernelsToString<Kernels>());
+
+    INFO("PrecisionT, ParamT = " << PrecisionToName<PrecisionT>::value << ", "
+                                 << PrecisionToName<ParamT>::value);
+
+    const auto all_wires = crateAllWires(num_qubits, gate_op, true);
+    for (const auto &wires : all_wires) {
+        const auto params = createParams<ParamT>(gate_op);
+        const auto gate_name = lookup(gate_names, gate_op);
+        DYNAMIC_SECTION(
+            "Test gate "
+            << gate_name
+            << " with inverse = false") { // Test with inverse = false
+            const auto results = Util::tuple_to_array(
+                applyGateForImplemetingKernels<gate_op, PrecisionT, ParamT,
+                                               Kernels>(
+                    ini, num_qubits, wires, false, params,
+                    std::make_index_sequence<length<Kernels>()>()));
+
+            for (size_t i = 0; i < results.size() - 1; i++) {
+                REQUIRE(results[i] ==
+                        PLApprox(results[i + 1])
+                            .margin(static_cast<PrecisionT>(1e-5)));
```

Comment:
Incorrect use of `PLApprox::margin`. You can reproduce the error with `clang-diagnostic-error`.

Reply:
Ah! I didn't know ab this :slightly_smiling_face: